src/socket/socket_ref.rs

```diff
@@ -0,0 +1,83 @@
+use core::ops::{Deref, DerefMut};
+use socket::{FromSocket, RawSocket, Socket, TcpSocket, UdpSocket, SocketHandle};
+
+/// A trait for tracking a socket usage sessions.
+///
+/// Allows implementation of custom on-drop logic.
+pub trait SocketSession {
+    fn finish(&mut self) {}
+}
+
+impl<'a, 'b> SocketSession for RawSocket<'a, 'b> {}
+impl<'a, 'b> SocketSession for UdpSocket<'a, 'b> {}
+impl<'a> SocketSession for TcpSocket<'a> {}
+
+impl<'a, 'b> SocketSession for Socket<'a, 'b> {
```

Comment:
We don't (and can't, Rust enums are not extensible) support new socket types defined in external crates, so the entire SocketSession logic can simply be folded into the SocketRef Drop impl.

Reply:
`SocketRef` is generic over socket type, we can't put socket-specific cleanup code there. Besides, I'm going to extend `SocketSession` with an associated type that holds socket's state, something like https://github.com/batonius/smoltcp/blob/packet_dispatch/src/socket/tracker.rs#L11 .